bothan-core/src/registry.rs

```diff
@@ -0,0 +1,150 @@
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::marker::PhantomData;
+
+use crate::registry::signal::Signal;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+pub mod post_processor;
+pub mod processor;
+pub mod signal;
+pub mod source;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Invalid;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Valid;
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Registry<State = Invalid> {
+    #[serde(flatten)]
+    inner: HashMap<String, Signal>,
+    #[serde(skip_serializing, default)]
+    _state: PhantomData<State>,
+}
+
+#[derive(Clone, Debug, Error, PartialEq)]
+pub enum ValidationError {
+    #[error("Circular dependency detected")]
+    CircularDependency,
+
+    #[error("Missing dependency")]
+    MissingDependency,
+}
+
+impl Registry<Invalid> {
+    pub fn validate(self) -> Result<Registry<Valid>, ValidationError> {
```

Comment:
I think DFS while maintaining the list of visited nodes will result in a lot simpler code.

Reply:
fixed